### PR TITLE
Filtering for specific frames for simple files representations

### PR DESCRIPTION
**Pull request by kalisp**

### CODE REVIEW

File: client/ayon_core/pipeline/farm/pyblish_functions.py [hunk 1]
```diff
@@ -295,11 +295,17 @@ def _add_review_families(families):
...
+def prepare_representations(
+    skeleton_data,
+    exp_files,
+    anatomy,
+    aov_filter,
+    skip_integration_repre_list,
+    do_not_add_review,
+    context,
+    color_managed_plugin,
+    frames_to_render
```
Comment (BigRoy) [inline client/ayon_core/pipeline/farm/pyblish_functions.py:307]: Should we make this new key optional for better backwards compatibility (if anyone was using it outside of us?)?

Comment (iLLiCiTiT) [reply client/ayon_core/pipeline/farm/pyblish_functions.py:307]: I 100% agree.

File: client/ayon_core/pipeline/farm/pyblish_functions.py [hunk 2]
```diff
@@ -325,6 +333,8 @@ def prepare_representations(skeleton_data, exp_files, anatomy, aov_filter,
 
     log = Logger.get_logger("farm_publishing")
 
+    frames_to_render = _get_real_frames_to_render(frames_to_render)
```
Comment (iLLiCiTiT) [inline client/ayon_core/pipeline/farm/pyblish_functions.py:336]: This has to change accordingly:
```python
if frames_to_render is not None:
    frames_to_render = _get_real_frames_to_render(frames_to_render)
else:
    frame_start = int(skeleton_data.get("frameStartHandle"))
    frame_end = int(skeleton_data.get("frameEndHandle"))
    frames_to_render = list(range(frame_start, frame_end + 1))
    
```

File: client/ayon_core/pipeline/farm/pyblish_functions.py [hunk 3]
```diff
@@ -453,6 +469,50 @@ def prepare_representations(skeleton_data, exp_files, anatomy, aov_filter,
...
+    """
+    frames_to_render = []
+    for frame in frames.split(","):
+        if "-" in frame:
+            splitted = frame.split("-")
+            frames_to_render.extend(
+                range(int(splitted[0]), int(splitted[1])+1))
+        else:
+            frames_to_render.append(frame)
+    return [str(frame_to_render) for frame_to_render in frames_to_render]
```
Comment (iLLiCiTiT) [inline client/ayon_core/pipeline/farm/pyblish_functions.py:485]: Why converted to string?
```suggestion
    return [frame_to_render for frame_to_render in frames_to_render]
```

File: client/ayon_core/pipeline/farm/pyblish_functions.py [hunk 4]
```diff
@@ -453,6 +469,50 @@ def prepare_representations(skeleton_data, exp_files, anatomy, aov_filter,
...
+    Artists could want to selectively render only particular frames
+    """
+    frames_to_render = []
+    for frame in frames.split(","):
+        if "-" in frame:
+            splitted = frame.split("-")
+            frames_to_render.extend(
+                range(int(splitted[0]), int(splitted[1])+1))
+        else:
+            frames_to_render.append(frame)
```
Comment (iLLiCiTiT) [inline client/ayon_core/pipeline/farm/pyblish_functions.py:484]:
```suggestion
            frames_to_render.append(int(frame))
```

File: client/ayon_core/pipeline/farm/pyblish_functions.py [hunk 5]
```diff
@@ -361,18 +371,21 @@ def prepare_representations(skeleton_data, exp_files, anatomy, aov_filter,
                 " This may cause issues on farm."
             ).format(staging))
 
-        frame_start = int(skeleton_data.get("frameStartHandle"))
+        frame_start = int(frames_to_render[0])
+        frame_end = int(frames_to_render[-1])
```
Comment (iLLiCiTiT) [inline client/ayon_core/pipeline/farm/pyblish_functions.py:375]:
```suggestion
        frame_start = frames_to_render[0]
        frame_end = frames_to_render[-1]
```

File: client/ayon_core/pipeline/farm/pyblish_functions.py [hunk 6]
```diff
@@ -453,6 +469,50 @@ def prepare_representations(skeleton_data, exp_files, anatomy, aov_filter,
...
+    """
+    files = [os.path.basename(f) for f in list(collection)]
+    file_name, extracted_frame = list(collect_frames(files).items())[0]
+    if extracted_frame:
+        found_frame_pattern_length = len(extracted_frame)
+        normalized_frames_to_render = set()
+        for frame_to_render in frames_to_render:
+            normalized_frames_to_render.add(
+                str(frame_to_render).zfill(found_frame_pattern_length)
+            )
```
Comment (iLLiCiTiT) [inline client/ayon_core/pipeline/farm/pyblish_functions.py:504]:
```suggestion
        normalized_frames_to_render = {
            str(frame_to_render).zfill(found_frame_pattern_length)
            for frame_to_render in frames_to_render
        }
```

File: client/ayon_core/pipeline/farm/pyblish_functions.py [hunk 7]
```diff
@@ -453,6 +469,50 @@ def prepare_representations(skeleton_data, exp_files, anatomy, aov_filter,
...
+    """Use expected files based on real frames_to_render.
+
+    Artists might explicitly set frames they want to render via Publisher UI.
+    This uses this value to filter out files
+    Args:
+        frames_to_render (list): of str '1001'
+    """
+    files = [os.path.basename(f) for f in list(collection)]
+    file_name, extracted_frame = list(collect_frames(files).items())[0]
+    if extracted_frame:
```
Comment (iLLiCiTiT) [inline client/ayon_core/pipeline/farm/pyblish_functions.py:498]: I would rever this logic...
```python
if not extracted_frame:
    return files
...
```

File: client/ayon_core/pipeline/farm/pyblish_functions.py [hunk 8]
```diff
@@ -453,6 +475,54 @@ def prepare_representations(skeleton_data, exp_files, anatomy, aov_filter,
...
+        )
+
+    filtered_files = []
+    for file_name in files:
+        if any(frame in file_name
+               for frame in normalized_frames_to_render):
+            filtered_files.append(file_name)
+
+    files = filtered_files
+    return files
```
Comment (iLLiCiTiT) [inline client/ayon_core/pipeline/farm/pyblish_functions.py:523]: 2 variables defined just to return this.
```suggestion
    return [
        file_name
        for file_name in files
        if any(frame in file_name
               for frame in normalized_frames_to_render):
    ]
```

File: client/ayon_core/pipeline/farm/pyblish_functions.py [hunk 9]
```diff
@@ -453,6 +475,54 @@ def prepare_representations(skeleton_data, exp_files, anatomy, aov_filter,
...
+
+    if not extracted_frame:
+        return files
+
+    found_frame_pattern_length = len(extracted_frame)
+    normalized_frames_to_render = set()
+    for frame_to_render in frames_to_render:
+        normalized_frames_to_render.add(
+            str(frame_to_render).zfill(found_frame_pattern_length)
+        )
```
Comment (iLLiCiTiT) [inline client/ayon_core/pipeline/farm/pyblish_functions.py:514]: Much faster than adding it one by one
```suggestion
    normalized_frames_to_render = {
        str(frame_to_render).zfill(found_frame_pattern_length)
        for frame_to_render in frames_to_render
    }
```

**Review by iLLiCiTiT — APPROVED** [review]

**Pull request merged**